Vk Pixel History: per fragment data

apazylbe opened the pull request:
- For now only per fragment primitive ID, and colour values, not depth
information.
- Adds two callbacks: first: to get shader output, post modification value, and primitive value; second: to do an occlusion query for primitives where some fragments were discarded  
- This still largely does not take into account multiple subpasses

apazylbe committed:
Vk Pixel History: per fragment data

- For now only per fragment primitive ID, and colour values, not depth
information.

baldurk commented on renderdoc/CMakeLists.txt:
```diff
@@ -383,6 +383,7 @@ set(data
     data/glsl/minmaxtile.comp
     data/glsl/quadresolve.frag
     data/glsl/quadwrite.frag
+    data/glsl/pixelhistory.frag
```
We might need more shaders than just this one for pixel history by the end, so I think a better name would be `pixelhistory_primid.frag`.

apazylbe replied:
Renamed.

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -142,6 +155,23 @@ struct PixelHistoryShaderCache
     return m_FixedColFS;
   }
 
+  VkShaderModule GetPrimitiveIdShader()
```
The vulkan shader cache already creates a `VkShaderModule` for each builtin shader. Since there's no customisation here (i.e. patching the SPIR-V or anything) could you use `GetBuiltinModule` instead of `GetBuiltinBlob`?

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -1410,6 +1441,671 @@ struct TestsFailedCallback : public VulkanPixelHistoryCallback
+        m_pDriver, cmd, VulkanRenderState::PipelineBinding::BindGraphics, false);
+  }
+
+  VkPipeline CreatePipeline(ResourceId pipe, uint32_t eid, bool dynamicScissor)
```
Not needed for this PR, but in future - do you think it would be possible and worthwhile to share some of this logic between callbacks? either with static helper functions in this file or by deriving from a common parent class?

apazylbe replied:
In the new commit I added common methods to the parent class of all callbacks:
- one to create a pipeline create info that can be further tweaked
- one to create renderpass with DS attachment, and optionally replacing the format of the colour attachment
- one to create framebuffer with DS attachmnet, and optionally replacing the color attachment view

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -1410,6 +1441,671 @@ struct TestsFailedCallback : public VulkanPixelHistoryCallback
+    VkPipeline newPipe =
+        CreatePipeline(state.graphics.pipeline, eid, p.dynamicStates[VkDynamicScissor]);
+    state.graphics.pipeline = GetResID(newPipe);
+    state.BindPipeline(m_pDriver, cmd, VulkanRenderState::PipelineBinding::BindGraphics, false);
```
`PipelineBinding` is a normal enum so the extra scope here is unnecessary.

apazylbe replied:
Fixed

baldurk commented on renderdoc/driver/vulkan/vk_pixelhistory.cpp:
```diff
@@ -1410,6 +1441,671 @@ struct TestsFailedCallback : public VulkanPixelHistoryCallback
+    // For every fragment except the last one.
+    // The framebuffer should have the values that it had before this event.
+    // Should we use the original framebuffer? Or the new framebuffer?
+    // If we use the new one, it won't have the
```
This comment seems cut off?

apazylbe replied:
Oops, sorry. Removed now. That was me thinking out in code comments.

apazylbe committed:
Vk Pixel History: refactoring, per fragment depth

- Refactor to use common funcs
- Add per fragment depth data
- Rename pixelhistory.frag to pixelhistory_primid.frag
- Fix how some per fragment pipelines are created

For per fragment data we create 3 pipelines
1) Post modification pipeline
changes the stencil state and scissors around the target pixel
2) Shader output pipeline
in addition to above, disables rasterization discard, depth bounds test,
culling and turns off blending
3) Primitive ID pipe
in addition to above, disables depth test and depth write, and changes
the fragment shader to only output primitive ID
